Datastore/src/Operation.php

```diff
@@ -871,4 +961,275 @@ private function sortEntities(array $entities, array $keys)
+                    ? Direction::ASCENDING
+                    : Direction::DESCENDING;
+            }
+        }
+
+        if (isset($query['filter'])) {
+            $query['filter'] = $this->convertFilterProps($query['filter']);
+        }
+
+        if (isset($query['limit']) && !is_array($query['limit'])) {
+            $query['limit'] = [
+                'value' => $query['limit']
+            ];
+        }
+
+        return $query;
+    }
+
+    /**
+     * Parse query into API compatible values
+     *
+     * @param array $gqlQuery
+     * @return array
+     */
+    private function parseGqlQuery(array $gqlQuery)
+    {
+        if (isset($gqlQuery['namedBindings'])) {
+            foreach ($gqlQuery['namedBindings'] as $name => &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        if (isset($gqlQuery['positionalBindings'])) {
+            foreach ($gqlQuery['positionalBindings'] as &$binding) {
+                if (!isset($binding['value'])) {
+                    continue;
+                }
+
+                $binding = $this->prepareQueryBinding($binding);
+            }
+        }
+
+        return $gqlQuery;
+    }
+
+    /**
+     * Convert Query filters to an API-compatible value.
+     *
+     * @param array $filter The input filter data
+     * @return array
+     */
+    private function convertFilterProps(array $filter)
+    {
+        if (isset($filter['propertyFilter'])) {
+            $operator = $filter['propertyFilter']['op'];
+
+            try {
+                if (is_int($operator)) {
+                    // verify that the operator, given as enum value, exists.
+                    PropertyFilterOperator::name($operator);
+                } else {
+                    // convert the operator, given as a string, to a grpc value.
+                    $operator = PropertyFilterOperator::value($operator);
+                }
+            } catch (\UnexpectedValueException $e) {
+                throw new \InvalidArgumentException($e->getMessage());
+            }
+
+            $filter['propertyFilter']['op'] = $operator;
+        }
+
+        if (isset($filter['compositeFilter'])) {
+            if ($filter['compositeFilter']['op'] == 'AND') {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::PBAND;
+            } elseif ($filter['compositeFilter']['op'] == 'OR') {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::PBOR;
+            } else {
+                $filter['compositeFilter']['op'] = CompositeFilterOperator::OPERATOR_UNSPECIFIED;
+            }
+            foreach ($filter['compositeFilter']['filters'] as &$nested) {
+                $nested = $this->convertFilterProps($nested);
+            }
+        }
+
+        return $filter;
+    }
```

Comment:
Shouldn't this logic be in a `Serializer` class? This makes me a bit nervous, because changes to these parameters will break the API... 

Is there a generic way we can do this logic? For instance, `Message::serializeToJsonString` will handle string enums (e.g. `"AND"` and `"OR"`)